app/internal/page/donor/check_your_lpa_test.go

```diff
@@ -82,41 +81,150 @@ func TestPostCheckYourLpaDigitalCertificateProvider(t *testing.T) {
 			donorStore.
 				On("Put", r.Context(), updatedLpa).
 				Return(nil)
 
-			shareCodeSender := newMockShareCodeSender(t)
-			shareCodeSender.
-				On("SendCertificateProvider", r.Context(), notify.CertificateProviderInviteEmail, testAppData, true, updatedLpa).
+			err := CheckYourLpa(nil, donorStore, shareCodeSender, nil, nil)(testAppData, w, r, lpa)
+			resp := w.Result()
+
+			assert.Nil(t, err)
+			assert.Equal(t, http.StatusFound, resp.StatusCode)
+			assert.Equal(t, page.Paths.LpaDetailsSaved.Format("lpa-id")+"?firstCheck=1", resp.Header.Get("Location"))
+		})
+	}
+}
+
+func TestPostCheckYourLpaDigitalCertificateProviderOnSubsequentChecks(t *testing.T) {
+	testCases := map[string]struct {
+		certificateProviderDetailsTaskState actor.TaskState
+		expectedTemplateId                  notify.TemplateId
+		expectedSms                         notify.Sms
+	}{
+		"cp not started": {
+			certificateProviderDetailsTaskState: actor.TaskNotStarted,
+			expectedTemplateId:                  notify.CertificateProviderDigitalLpaDetailsChangedNotSeenLpaSMS,
+			expectedSms: notify.Sms{
+				PhoneNumber: "07700900000",
+				TemplateID:  "template-id",
+				Personalisation: map[string]string{
+					"donorFullName": "Teneil Throssell",
+					"lpaType":       "property and affairs",
+				},
+			},
+		},
+		"cp in progress": {
+			certificateProviderDetailsTaskState: actor.TaskInProgress,
+			expectedTemplateId:                  notify.CertificateProviderDigitalLpaDetailsChangedSeenLpaSMS,
+			expectedSms: notify.Sms{
+				PhoneNumber: "07700900000",
+				TemplateID:  "template-id",
+				Personalisation: map[string]string{
+					"donorFullNamePossessive": "Teneil Throssell’s",
+					"lpaType":                 "property and affairs",
+					"lpaId":                   "lpa-id",
+					"donorFirstNames":         "Teneil",
+				},
+			},
+		},
+		"cp completed": {
+			certificateProviderDetailsTaskState: actor.TaskCompleted,
+			expectedTemplateId:                  notify.CertificateProviderDigitalLpaDetailsChangedSeenLpaSMS,
+			expectedSms: notify.Sms{
+				PhoneNumber: "07700900000",
+				TemplateID:  "template-id",
+				Personalisation: map[string]string{
+					"donorFullNamePossessive": "Teneil Throssell’s",
+					"lpaType":                 "property and affairs",
+					"lpaId":                   "lpa-id",
+					"donorFirstNames":         "Teneil",
+				},
+			},
+		},
+	}
+
+	for testName, tc := range testCases {
+		t.Run(testName, func(t *testing.T) {
+			form := url.Values{
+				"checked-and-happy": {"1"},
+			}
+
+			w := httptest.NewRecorder()
+			r, _ := http.NewRequest(http.MethodPost, "/", strings.NewReader(form.Encode()))
+			r.Header.Add("Content-Type", page.FormUrlEncoded)
+
+			localizer := newMockLocalizer(t)
+			localizer.
+				On("T", "pfaLegalTerm").
+				Return("property and affairs")
+
+			if !tc.certificateProviderDetailsTaskState.NotStarted() {
+				localizer.
+					On("Possessive", "Teneil Throssell").
+					Return("Teneil Throssell’s")
```

Comment:
You can add a `.Maybe()` to calls that might not happen, which would be ok here because this value must flow through for other checks to work. I think it is sometimes neater than an `if`

Reply:
Gamechanger! Thanks for that, will update.